x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,739 @@
+	if k, found := m.canonical[DeterministicStringify(pk)]; found {
+		power, found := m.inner[k]
+		if !found {
+			panic("found canonical key but key not present in inner map")
+		}
+		return power, true
+	}
+	return -1, false
+}
+
+// KeyToLastUpdateMemo is an implementation of a map from a public key to a
+// memo storing the last update.
+// It is used because public keys are not comparable in Go, so they cannot be
+// used as builtin map keys.
+type KeyToLastUpdateMemo struct {
+	inner     map[tmprotocrypto.PublicKey]providertypes.LastUpdateMemo
+	canonical map[string]tmprotocrypto.PublicKey
+}
+
+func MakeKeyToLastUpdateMemo() KeyToLastUpdateMemo {
+	return KeyToLastUpdateMemo{
+		inner:     map[tmprotocrypto.PublicKey]providertypes.LastUpdateMemo{},
+		canonical: map[string]tmprotocrypto.PublicKey{},
+	}
+}
+
+// Set is a typical map set operation.
+func (m *KeyToLastUpdateMemo) Set(pk tmprotocrypto.PublicKey, memo providertypes.LastUpdateMemo) {
+	stringified := DeterministicStringify(pk)
+	if k, found := m.canonical[stringified]; found {
+		pk = k
+	} else {
+		m.canonical[stringified] = pk
+	}
+	m.inner[pk] = memo
+}
+
+// Get is a typical map get operation.
+func (m *KeyToLastUpdateMemo) Get(pk tmprotocrypto.PublicKey) (providertypes.LastUpdateMemo, bool) {
+	if k, found := m.canonical[DeterministicStringify(pk)]; found {
+		memo, found := m.inner[k]
+		if !found {
+			panic("found canonical key but key not present in inner map")
+		}
+		return memo, true
+	}
+	return providertypes.LastUpdateMemo{}, false
+}
+
+// KeySet is an implementation of a set of public keys.
+// It is used because public keys are not comparable in Go
+// so they cannot be used as builtin map keys.
+type KeySet struct {
+	inner     map[tmprotocrypto.PublicKey]struct{}
+	canonical map[string]tmprotocrypto.PublicKey
+}
+
+func MakeKeySet() KeySet {
+	return KeySet{
+		inner:     map[tmprotocrypto.PublicKey]struct{}{},
+		canonical: map[string]tmprotocrypto.PublicKey{},
+	}
+}
+
+// Add is a typical set add operation.
+func (s *KeySet) Add(pk tmprotocrypto.PublicKey) {
+	stringified := DeterministicStringify(pk)
+	if k, found := s.canonical[stringified]; found {
+		pk = k
+	} else {
+		s.canonical[stringified] = pk
+	}
+	s.inner[pk] = struct{}{}
+}
+
+// Add is a typical set has/contains operation.
+func (s *KeySet) Has(pk tmprotocrypto.PublicKey) bool {
+	_, found := s.canonical[DeterministicStringify(pk)]
+	return found
+}
+
+// getProviderKeysForUpdate returns the all the provider keys of validators for which
+// an update must be sent. An update must be sent to the consumer if, either:
+//  1. The consumer has the validator in its active set AND the assigned key for the
+//     validator has changed since the last update.
+//  2. The voting power of the validator has changed, and the validator is in the active
+//     set.
+func (ka *KeyAssignment) getProviderKeysForUpdate(stakingUpdates KeyToPower) KeySet {
+
+	ret := MakeKeySet()
+
```

Comment:
simple way to put it "find validators in which we need to update their key on a consumer through a vsc update"